compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp

```diff
@@ -0,0 +1,926 @@
+    }
+
+    DistributionLayout *resultLattice = getLatticeElement(result);
+    resultLattices.push_back(resultLattice);
+  }
+
+  // Exit early on operations with no results.
+  if (resultLattices.empty()) {
+    return;
+  }
+
+  auto changeFunc = [&](DistributionLayout *lattice, ChangeResult changed) {
+    this->propagateIfChanged(lattice, changed);
+  };
+
+  propagationTransferFunction(op, operandLattices, resultLattices, changeFunc);
+}
+
+void PropagateLayout::visitRegionSuccessors(RegionBranchOpInterface branch,
+                                            RegionBranchPoint branchPoint,
+                                            OperandRange operands) {
+  SmallVector<RegionSuccessor> successors;
+  branch.getSuccessorRegions(branchPoint, successors);
+  for (RegionSuccessor successor : successors) {
+    ValueRange inputs = successor.getSuccessorInputs();
+
+    // Get vector layouts for forwarded operands.
+    SmallVector<const DistributionLayout *> forwardedLattices;
+    for (Value operand : operands) {
+      if (isa<VectorType>(operand.getType())) {
+        forwardedLattices.push_back(getLatticeElement(operand));
+      }
+    }
+
+    // Get vector layouts for input operands.
+    SmallVector<DistributionLayout *> inputLattices;
+    for (Value operand : inputs) {
+      if (isa<VectorType>(operand.getType())) {
+        inputLattices.push_back(getLatticeElement(operand));
+      }
+    }
+
+    // Both should have same number of vector operands.
+    assert(forwardedLattices.size() == inputLattices.size() &&
+           "Number of forwarded operands and inputs should match");
+
+    // Propagate the layouts.
+    for (auto [forwardedLattice, inputLattice] :
+         llvm::zip(forwardedLattices, inputLattices)) {
+      inputLattice->resolve(forwardedLattice);
+    }
+  }
+}
+
+DistributionLayout *PropagateLayout::getLatticeElement(Value val) {
+  // Add dependency of operation on the analysis state.
+  assert(isa<VectorType>(val.getType()) && "Lattice value should be a vector");
+  DistributionLayout *layout =
+      DataFlowAnalysis::getOrCreate<DistributionLayout>(val);
+  // Subscribe this analysis to updates of the lattice.
+  layout->subscribePropagation(this);
+  return layout;
+}
+
+/// ==========================================================================
+///        Enforce Layout
+/// ==========================================================================
+
+LogicalResult EnforceLayout::initialize(Operation *root) {
+  root->walk([&](Operation *traversed) { visitOperation(traversed); });
+  return success();
+}
+
+LogicalResult EnforceLayout::visit(ProgramPoint point) {
+  if (Operation *op = dyn_cast_or_null<Operation *>(point)) {
+    visitOperation(op);
+    return success();
+  }
+
+  // Do not expect anything else.
+  return failure();
+}
+
+void EnforceLayout::visitOperation(Operation *op) {
+  // Handle region branching control flow.
+  // TODO: Write more about what we are doing here.
```

Comment:
Just wondering what ops are covered by RegionBranchOpInterface and are not covered by RegionBranchTerminatorOpInterface + RegionBranchOpInterface ? Why do we need the first if statement (doesn't the second cover it?)

Reply:
They have different inputs. Only RegionBranchOpInterface has inputs as the init_args. With a yield, the inputs are the yielded arguments. Also, the region successors may be different.